devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp

```diff
@@ -303,75 +303,258 @@ bool WholeBodyDynamicsDevice::openEstimator(os::Searchable& config)
+            {
+                yWarning() << "Frame " << overrideContactFrames[i] << " not found in the model, discarding it";
+            }
+            else
+            {
+                overrideContactFramesIdx.push_back(idx);
+            }
+        }
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// The frames from a single submodel should be no more than 1 if the type is FULL_WRENCH, no more than 3 if the type is PURE_FORCE, and no more than 6 if the type is PURE_FORCE_WITH_KNOWN_DIRECTION.
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+        // For each submodel, we find the first suitable contact frame
+        // This is a n^2 algorithm, but given that is just used in
+        // configuration it should be ok
+        size_t nrOfSubModels = estimator.submodels().getNrOfSubModels();
+
+        //// We indicate with FRAME_INVALID_INDEX the fact that we still don't have a override contact for the given submodel
+        subModelIndex2OverrideContact.resize(overrideContactFramesIdx.size());
+
+        subModelVarSize.resize(nrOfSubModels);
+
+        for(size_t i=0; i < overrideContactFramesIdx.size(); i++)
         {
-            yError() << "wholeBodyDynamics : openDefaultContactFrames : missing default contact for submodel composed by the links: ";
-            const iDynTree::Traversal & subModelTraversal = estimator.submodels().getTraversal(subModelIdx);
-            for(iDynTree::TraversalIndex i = 0; i < (iDynTree::TraversalIndex) subModelTraversal.getNrOfVisitedLinks(); i++)
+            // We indicate with FRAME_INVALID_INDEX the fact that we still don't have a override contact for the given submodel
+            subModelIndex2OverrideContact[i].resize(nrOfSubModels,iDynTree::FRAME_INVALID_INDEX);
+
+            size_t subModelIdx = estimator.submodels().getSubModelOfFrame(estimator.model(),overrideContactFramesIdx[i]);
+
+            //check if adding the current contact to this subModel will increase the no of variable to more than 6
+            if(subModelVarSize[subModelIdx]+wrench[i].nrUnknownsInExtWrench > 6)
             {
-                iDynTree::LinkIndex linkIdx = subModelTraversal.getLink(i)->getIndex();
-                yError() << "wholeBodyDynamics : openDefaultContactFrames :" << estimator.model().getLinkName(linkIdx);
+                //subModelIndex2OverrideContact[subModelIdx].push_back(iDynTree::FRAME_INVALID_INDEX);
+                yWarning() << "subModel no " << subModelIdx << " has the maximum number of variables (6). The frame " << overrideContactFrames[i] << " will not be added to it.";
             }
+            else
+            {
+                //add the contact frame
+                //subModelIndex2OverrideContact[subModelIdx].push_back(overrideContactFramesIdx[i]);
```

Comment:
the `push_back` will be wrong because the order of the contact frames can be arbitrary and not depend on how the submodels are ordered. Could you please remove this commented line? and add a line of comment here as to why we use array-indexing instead of  push_back?

Reply:
You're right. done in https://github.com/robotology/whole-body-estimators/pull/48/commits/edcd509dad868124ec98e627ecd260af425c8ca1